### DISCUSSION TITLE
Create vertices "on the fly"

### DISCUSSION THREAD

**Discussion by alexandernst** [clientIO/joint#2301] [Ideas]
I was playing around with a tool for circuit design and the way links are created in that tool is (imho) very helpful. When links are being created, vertices can be dynamically added by making a right click while still dragging the link.

Visual example: https://www.youtube.com/watch?v=zb5OJxtPjvw&t=355s

Would this be possible with JointJS?

**Comment by kumilingus** [comment]
This seems to be a useful feature.
For example, the [Connect](https://github.com/clientIO/joint/blob/master/src/linkTools/Connect.mjs) tool can be modified to behave as described above.

Here's an [example](https://jsfiddle.net/kumilingus/mekpdjof/).

**Reply by alexandernst** [reply]
This is exactly what I was looking for!! Thank you!

**Reply by alexandernst** [reply]
I have been doing some tests with this and while the code works exactly as I asked for, I didn't account for the fact that this won't be usable by people with trackpads (there is not way to drag and right-click at the same time using a trackpad).

So... I'm gonna ask for something a little bit more complex (a-la-brainstorming).

Is it possible to do the following?:

* implement what I asked (which would be very helpful for people with regular mouse)

**and also, at the same time**

* make it possible to start creating links by clicking on a port and then:
  * if a click on the blank area is performed, create a vertex
  * if a click on another port is performed, finish creating the link
  * if <kbd>Esc</kbd>  is hit, abort the creation of the link

That would mean that JJ would need to be able to differentiate between a single click (pointer down & up) and click&drag in order to chose which behaviour to trigger.

**Reply by kumilingus** [reply]
That's already possible. You can set the `magnetThreshold` of the paper to `onleave` and listen for `element:magnet:pointerclick` to start the link creation by clicking. You should listen to the `mousedown` event of the `document` (in the `capture` phase - to prevent propagation) to see what DOM node was clicked and decide whether to add a new vertex, connect the link to a port, or cancel the action.

We can prepare an example for one of the upcoming [Demo Wednesday](https://github.com/clientIO/joint/discussions?discussions_q=is%3Aopen+label%3A%22Demo+Wednesday%22).

**Reply by alexandernst** [reply]
I need to try this, will get back with my results! :)

Yes please! A demo would be very helpful 🙏

**Reply by alexandernst** [reply]
@kumilingus demo-wednesday-ping ^ 🥺

**Reply by kumilingus** [reply]
@alexandernst after the `v4` release. Feel free to ping me again.

**Reply by kumilingus** [reply]
@alexandernst would something like [this example](https://codepen.io/jointjs/pen/poYGdLz/35ded2458fad9d89c87b893ccc8aa237?editors=0010) work for you?

**Reply by alexandernst** [reply]
Hi Roman! I studied the code and this is almost what I was expecting! The only thing that I'm not really sure how to do (if it's possible) is to trigger the vertex creation (with a right click) while physically dragging the arrow of the link. As so:

1. Left-click on a port (don't release the button)
2. Start dragging
3. right click <-- at this point a new vertex should be created

I made some changes to the code in order to test some stuff, but I wasn't able to make JJ to prevent the default action to get triggered (show the contextual menu of the web browser).

**Reply by kumilingus** [reply]
I am not able to test this code properly because I have no mouse at the moment. Could you check please?

```js
const paper = new joint.dia.Paper({ /* ... */ });

paper.pointerup = function(evt) {
    const view = this.eventData(evt).sourceView;
    if (view) {
        const { linkView, action } = view.eventData(evt);
        if (action === 'magnet' && linkView && evt.button === 2) {
            const normalizedEvt = joint.util.normalizeEvent(evt);
            const { x, y } = this.clientToLocalPoint({ x: normalizedEvt.clientX, y: normalizedEvt.clientY });
            linkView.model.vertices(linkView.model.vertices().concat({ x, y }));
            return;
        }
    }
    joint.dia.Paper.prototype.pointerup.call(this, evt);
}
```

**Reply by alexandernst** [reply]
The context menu is still triggering (and thus making it impossible to keep interacting with the paper / link). I tried calling `preventDefault()` and `stopPropagation()`, but it looks like the context menu is launched before the execution control has called the `pointerup` fn. Since JJ is able to actually prevent the context menu, I'm 100% sure I'm just missing something, but I can't figure out what is it.

**Reply by kumilingus** [reply]
Is [this](https://stackoverflow.com/questions/45510938/how-to-disable-oncontextmenu-while-rightclick-dragging) helpful?

**Reply by alexandernst** [reply]
Yes!! I was actually trying to call `preventDefault()` using JJ instead of native event listener (more on that in a moment).

So, here is the final version: https://codepen.io/alexandernst/pen/KKEJooO?editors=0011

I'm also pasting the code here, just in case something happens with the codepen link

```js
const shapes = {
    ...joint.shapes,
};

const graph = new joint.dia.Graph({}, { cellNamespace: shapes });
const paper = new joint.dia.Paper({
    linkPinning: false,
    el: document.getElementById('paper'),
    width: 800,
    height: 600,
    model: graph,
    cellViewNamespace: shapes,
    interactive: { linkMove: false },
    magnetThreshold: 'onleave',
});

const r1 = new joint.shapes.standard.Rectangle({
    position: { x: 100, y: 100 },
    size: { width: 100, height: 40 },
    attrs: {
        root: { magnet: false },
        body: { fill: 'gray' }, label: { fill: 'white' }
    },
    ports: {
        groups: {
            in: {
                position: { name: 'top' },
                attrs: {
                    circle: { r: 8, magnet: true, stroke: 'gray', fill: 'lightgray', strokeWidth: 2 }
                }
            },
            out: {
                position: { name: 'bottom' },
                attrs: {
                    circle: { r: 8, magnet: true, stroke: 'gray', fill: 'lightgray', strokeWidth: 2 }
                }
            }
        },
        items: [{ id: 'in1', group: 'in' }, { id: 'out1', group: 'out' }]

    }
});

const r2 = r1.clone().translate(300, 0);


graph.addCells([r1, r2]);


////////// this lets you create vertex by dragging and right-clicking ///////////////////
document.addEventListener('contextmenu', (evt) => evt.preventDefault());

paper.pointerup = function(evt) {
    const view = this.eventData(evt).sourceView;
    if (view) {
        const { linkView, action } = view.eventData(evt);
        if (action === 'magnet' && linkView && evt.button === 2) {
            const normalizedEvt = joint.util.normalizeEvent(evt);
            const { x, y } = this.clientToLocalPoint({ x: normalizedEvt.clientX, y: normalizedEvt.clientY });
            linkView.model.vertices(linkView.model.vertices().concat({ x, y }));
            return;
        }
    }
    joint.dia.Paper.prototype.pointerup.call(this, evt);
}

////////// this lets you create vertex by clicking ////////////////
let linking = false;
paper.on('element:magnet:pointerclick', function(elementView, evt, magnet, x, y) {
    if (linking) {
        linking = false;
        return;
    }

    graph.startBatch('add-link');

    linking = true;

    const sourceId = elementView.model.id;
    const sourcePort = elementView.findAttribute('port', magnet);

    const link = new joint.shapes.standard.Link({
        source: { id: sourceId, port: sourcePort },
        target: { x, y },
        attrs: {
            root: { pointerEvents: 'none' },
            line: { targetMarker: null }
        },
    });

    graph.addCell(link);

    const done = () => {
        link.attr('root/pointerEvents', 'auto');
        link.attr('line/targetMarker', { type: 'path', d: 'M 10 -5 0 0 10 5 Z' });
        graph.stopBatch('add-link');
        document.removeEventListener('pointermove', onPointermove);
        document.removeEventListener('pointerup', onPointerup);
        document.removeEventListener('keydown', onEscape);
    };

    const onPointermove = (evt) => {
        const p = paper.clientToLocalPoint({ x: evt.clientX, y: evt.clientY });
        link.target(p);
    };

    const onPointerup = (evt) => {
        if (evt.button === 2) {
            link.vertices(link.vertices().slice(0, -1));
            return;
        }
        const view = paper.findView(evt.target);
        if (view) {
            const targetPort = view.findAttribute('port', evt.target);
            if (targetPort && (view.model.id !== sourceId || sourcePort !== targetPort)) {
                link.target({ id: view.model.id, port: targetPort });
                done();
                return;
            }
        } else {
            const p = paper.clientToLocalPoint({ x: evt.clientX, y: evt.clientY });
            link.vertices(link.vertices().concat(p.toJSON()));
        }
    };

    const onEscape = (evt) => {
        if (evt.key === 'Escape') {
            link.remove();
            done();
        }
    };

    document.addEventListener('pointermove', onPointermove);
    document.addEventListener('pointerup', onPointerup);
    document.addEventListener('keydown', onEscape);
});
```

I red several times the [docs about events](https://resources.jointjs.com/docs/jointjs/v4.0/joint.html#dia.Paper.events) and I'm not sure why, but I was expecting `paper.on("contextmenu")` to behave the same way as `document.addEventListener("contextmenu")`. Maybe that would be a nice little feature to be added to JJ?

Also, related question: the very first time (after a page refresh) I try to create a link by dragging, the arrow on the head of the link doesn't show. Consecutive retries fix that and the arrow on the head of the link is shown. 
If I try to create a link by clicking (instead of dragging), the arrow on the head of the link isn't drawn at all (neither the first nor consecutive attempts). Why is that?

**Reply by kumilingus** [reply]
> I red several times the [docs about events](https://resources.jointjs.com/docs/jointjs/v4.0/joint.html#dia.Paper.events) and I'm not sure why, but I was expecting paper.on("contextmenu") to behave the same way as document.addEventListener("contextmenu"). Maybe that would be a nice little feature to be added to JJ?

The problem is that we undelegate all events on drag start and delegate them back after drag ends. This is to prevent events such  `mouseover` to be triggered while the user is dragging.
We could disable only specific events I believe, but a desired solution is to start using `pointer` events (not `mouse` and `touch`) and capture the target with `setPointerCapture()` ([example](https://jsfiddle.net/kumilingus/xLgeuhqy/)).

> Also, related question: the very first time (after a page refresh) I try to create a link by dragging, the arrow on the head of the link doesn't show. Consecutive retries fix that and the arrow on the head of the link is shown.
If I try to create a link by clicking (instead of dragging), the arrow on the head of the link isn't drawn at all (neither the first nor consecutive attempts). Why is that?

Do you mean in Safari? Then it is perhaps related to this [bug](https://bugs.webkit.org/show_bug.cgi?id=268376).

**Reply by alexandernst** [reply]
Ahh, yes, I remember! I should have known since I actually wrote some code related to that 😂. Bad memory, I guess.

Dang it! It's always safari. I did a quick test with a few different versions and wrote a comment in your bug report. Lets hope that they'll be more willing to fix it now that they know which version broke it. (git bisect-ing should be doable)

**Reply by kumilingus** [reply]
Well done & thanks!

_Pardon my ignorance, but how do you test various versions of Safari?_

**Reply by alexandernst** [reply]
I was on your boat 5 hours ago, no ignorance at all! 😂
Found this https://www.lambdatest.com/online-browser-testing and it actually worked quite well

**Reply by kumilingus** [reply]
Haha, thanks, indeed a nice tool.

**Reply by alexandernst** [reply]
I just noticed something strange. When creating links with a click (not by dragging from the source port), the `snapLinks / radius` property isn't honored. Why could that be? 🤔

**Reply by alexandernst** [reply]
It also doesn't honor `validateConnection`, but I think this is expected since the link is manually created. Is there some sort of `link_model.isValid()` I can call in order to check if the link is "valid" (according to `validateConnection`)?

**Reply by kumilingus** [reply]
The code I shared only keep the position of the arrowhead under the pointer. It does not run any validation nor snapping logic.
1. Solving the `validation` is easy, as you can have a separate validation function that is executed from the paper `validateConnection()` callback and from the code I shared with you.
2. To enable `snapLinks` you'd have to call `linkView.dragLinkStart`, `linkView.dragLink()` and `linkView.dragLinkEnd()` as shown in the other [example](https://jsfiddle.net/kumilingus/mekpdjof/) (this would call the `validateConnection()` logic as well). Unfortunately, there is no official API at this moment to control the link dragging programmatically.

**Reply by alexandernst** [reply]
I finally managed to allocate some time and dig into this (sigh... more than 6 months...).

As per my previous example, I'm using `document.addEventListener` to subscribe to the mouse down/move/up events, which are required to implement the behaviour I want. Your suggestion (call `dragLinkStart()`, `dragLink()` and `dragLinkEnd()`) made me wonder how would I actually do it. The problem is that these 3 methods expect to receive a `jj event` (let's call it that way 😂), but the only thing I can provide them with is a `native mouse event`, because that's what I'm getting from `document.addEventListener`.

I can't can swap the `document.addEventListener` with `paper.on` because there is no `pointermove` event in the paper (`blank:pointermove` will trigger only while the left button is pressed), so I'm not really sure how can I provide these methods with a `jj event`.

Is there a way I can get a `blank:pointermove` events from the paper without pressing the left button? Alternatively, is there a way I can "create" a `jj event`? Or maybe you can give me some other ideas how to tackle this problem?

**Reply by kumilingus** [reply]
I see. _JointJS_ event is just a wrapper for native events. This should work:
```ts
const jjEvent = new joint.mvc.$.Event(nativeEvent);
```

**Reply by alexandernst** [reply]
I think I managed to handle it! Thank you 🙏
